epidose/device/wps_scanner_d.sh

```diff
@@ -0,0 +1,95 @@
+#!/bin/sh
+#
+# Check for available networks and connect to the closest one using the WPS PBS
+#
+# Copyright 2020 Stefanos Georgiou
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+set -e
+
+export APP_NAME=wps_scanner_d
+
+# Location of suspend watchdod led
+SUSPEND_WATCHDOG_LED=/var/lock/epidose/suspend-watchdog-led
+
+# Pick up utility functions relative to the script's source code
+UTIL="$(dirname "$0")/util.sh"
+
+# Source common functionality (logging, WiFi)
+# shellcheck source=epidose/device/util.sh
+SERVER_URL="http://ha-server:5010" . "$UTIL"
+
+wps_connect()
+{
+  # Wait for the Wifi button to be pressed
+  run_python device_io --wifi-wait
+  wifi_acquire
+
+  network_status=$(ifconfig wlan0 | grep "inet ")
+
+  # Check if device is connected to any network
+  if [ -z "$network_status" ] ; then
+    log "Iniating WPS connection"
+    
+    # Connected to the network that supports WPS and has the lowest signal level
+    wpa_cli scan_results | grep WPS | sort -r -k3 | tail -1 | awk '{print $1}' | xargs wpa_cli -i wlan0 wps_pbc
+    log "Trying to connect to $(wpa_cli scan_results | grep WPS | sort -r -k3 | tail -1 | awk '{print $NF}')"
+    
+    # Obtain dynamic IP address
+    dhclient wlan0
+
+    # Suspend watchdog's LED activities
+    touch "$SUSPEND_WATCHDOG_LED"
+
+    check_network_again=$(ifconfig wlan0 | grep "inet ")
+    
+    # Check whether the connection is established and light up the LED accordingly
+    if [ -z "$check_network_again" ] ; then
+      # If failed to connect to a network, turn on red LED light
+      run_python device_io --red-on
+      log "Unable to connect to a network"
+      sleep 5
+      run_python device_io --red-off
+      exit_code=1
+    else
+      # If connected to a network, turn on green LED light
+      run_python device_io --green-on
+      log "Obtained network connection"
+      sleep 5
+      run_python device_io --green-off
+      exit_code=0
+    fi
+
+    # Release watchdog
+    rm "$SUSPEND_WATCHDOG_LED"
+
+    return $exit_code
+  fi
+}
+
+# Wait for WPS button to be pressed
+while : ; do
+  if ! wps_connect; then
+    log "Failed to connect after using the WPS PBS"
+  else
+    check_cuckoo_filter
+    if [ "$to_sleep" -eq 0 ] ; then
+      log "Updating filter"
+      try_to_get_filter_from_server
+    fi
+    #run_python check_interface_risk "$FILTER" || :
+  fi
+  wifi_release
```

Comment:
Try to place wifi_acquire and release in symmetric positions, making it easy to verify that these calls always match.

Reply:
Ok, now the wifi_acquire is done before the wps_connect function call and the wifi_release after the function.